libs/cardano-data/src/Data/Incremental.hs

```diff
@@ -0,0 +1,181 @@
+monoidInsert k v1 (MM m) = if v1 == mempty then MM (delete k m) else MM (insert k v1 m)
+{-# INLINEABLE monoidInsert #-}
+
+instance (Show k, Show v) => Show (MonoidMap k v) where
+  show (MM m) = show (Map.toList m)
+
+instance (Show (Diff v), Show k) => Show (Diff (MonoidMap k v)) where
+  show (Dm x) = show (Map.toList x)
+
+-- =========================================
+-- ILC instances
+
+-- | Monoidal maps have special properties, so they get their
+--   own instance (wrapped in the newtype).
+instance (Ord k, Eq v, ILC v, Monoid v) => ILC (MonoidMap k v) where
+  newtype Diff (MonoidMap k v) = Dm (Map k (MonoidD (Diff v)))
+  applyDiff mm (Dm md) = Map.foldlWithKey' accum mm md
+    where
+      accum :: MonoidMap k v -> k -> MonoidD (Diff v) -> MonoidMap k v
+      accum (MM ans) cred Del = MM (Map.delete cred ans)
+      accum ans cred (Comb dv) =
+        monoidInsertWith cred (applyDiff mempty dv) ans
+      accum ans cred (Write dv) = monoidInsert cred (applyDiff mempty dv) ans
+  {-# INLINEABLE applyDiff #-}
+  zero = Dm Map.empty
+  extend (Dm x) (Dm y) = Dm (Map.unionWith (<>) x y)
+
+deriving newtype instance (NFData k, NFData (Diff v)) => NFData (Diff (MonoidMap k v))
+
+-- | Normal map can only be deleted or updated so they use BinaryD
+instance Ord k => ILC (Map k v) where
+  newtype Diff (Map k v) = Dn (Map k (BinaryD v))
+  applyDiff m (Dn md) = Map.foldlWithKey' accum m md
+    where
+      accum ans k Omit = Map.delete k ans
+      accum ans k (Edit drep) = Map.insert k drep ans
+  {-# INLINEABLE applyDiff #-}
+  zero = Dn Map.empty
+  extend (Dn x) (Dn y) = Dn (Map.unionWith (<>) x y)
+
+instance (Show k, Show v) => Show (Diff (Map k v)) where
+  show (Dn x) = show (Map.toList x)
+
+deriving newtype instance (NFData k, NFData v) => NFData (Diff (Map k v))
+
+-- =================================================================
+-- helper functions for making binary derivatives
+
+-- | insert a change (MonoidD c) into a Map.
+--   Note that if we wrap the (result :: Map k (MonoidD c)) with the constructor 'Dn'
+--   Dn :: Map k (BinaryD v) -> Diff (Map k v)
+--   then we get Diff(Map k v)
+insertC ::
+  (Ord k, Monoid c) =>
+  k ->
+  MonoidD c ->
+  Map k (MonoidD c) ->
+  Map k (MonoidD c)
+insertC d m x = insertWith (<>) d m x
+
+-- | Split two maps, x and y, into three parts
+--   1) the key appears only in x
+--   2) the key appears in both x and y
+--   3) the key appears only in y
+--   Given three 'C'ontinuation style functions, reduce
+--   the three parts to a single value.
+inter3C ::
+  Ord k =>
+  a ->
+  Map k u ->
+  Map k v ->
+  (a -> k -> u -> a) ->
+  (a -> k -> (u, v) -> a) ->
+  (a -> k -> v -> a) ->
+  a
+inter3C ans0 m0 n0 c1 c2 c3 = go ans0 m0 n0
+  where
+    go ans Tip Tip = ans
+    go ans m Tip = (Map.foldlWithKey' c1 ans m)
+    go ans Tip n = (Map.foldlWithKey' c3 ans n)
+    go ans (Bin _ kx x l r) n = case Map.splitLookup kx n of
```

Comment:
Should we do the same for the two cases that use the Map.foldlWithKey',   note that function is not necessarily strict in ans?

Reply:
It wouldn't hurt adding `!acc` for other cases as well,but it is not really required, since `foldlWithKey'`  will force it. The last case however is the recursive call to `go`, which could build up a lot of thunks